### PR TITLE
fix: [DHIS2-11567] avoid using program.getOrganisationUnits due to high mem allocation (2.37)

### PR DESCRIPTION
**Pull request by ameenhere**
For a previous fix, the method program.getOrganisationUnits was used which is a very very heavy method and should not be used due to risk of high mem consumption. 
This PR aims at introducing a jdbc method which gets all the program associations without the high mem usage issue and replaces the program.getOrganisationUnits() usage with this jdbc method.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java [hunk 1]
```diff
@@ -730,11 +732,12 @@ private String validateProgramForEnrollment( Program program, Enrollment enrollm
                 " is a program without registration. An enrollment cannot be created into program without registration.";
         }
 
-        if ( program.getOrganisationUnits() != null && program.getOrganisationUnits().size() > 0 )
+        IdentifiableObjectAssociations programAssociations = programService
+            .getAllProgramOrganisationUnitsAssociations( Collections.singleton( program.getUid() ) );
+        if ( programAssociations.get( program.getUid() ) != null
+            && programAssociations.get( program.getUid() ).size() > 0 )
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:738]: Maybe `!CollectionUtils.isEmpty(programAssociations.get( program.getUid() ) )` is more coincise

**Review by enricocolasante — APPROVED** [review]
One minor comment

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/ProgramService.java [hunk 2]
```diff
@@ -159,5 +159,27 @@
...
+     * Get all the organisation unit associated for a set of program uids. This
+     * method uses jdbc to directly fetch the associated org unit uids for every
+     * program uid. This method returns all the associations irrespective of the
+     * sharing settings or org unit scopes.
+     *
+     * @param programUids A set of program uids
+     * @return A object of {@link IdentifiableObjectAssociations} containing
+     *         association for each programUid
+     */
+    IdentifiableObjectAssociations getAllProgramOrganisationUnitsAssociations( Set<String> programUids );
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/ProgramService.java:184]: Perhaps remove the "All" part from the method name. Sort of cleaner without. After all it is not returning all, it is limited by the given program UIDs.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java (hunk 1 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/enrollment/AbstractEnrollmentService.java:738]: Resolved!

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/ProgramService.java (hunk 2 above)
Comment (ameenhere) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/ProgramService.java:184]: Resolved. 
It was added to have a more visual difference from the other method which filters the "org units". This method fetches "All org unit associates for the specified programs".

**Pull request merged**